---
title: "ChIP_Seq_Analysis"
author: "Maximilian Greil"
date: "21 4 2021"
output:
  html_document:
    theme: journal
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

```{r}
suppressMessages(library(UpSetR))
suppressMessages(library(ChIPpeakAnno))
suppressMessages(library(TxDb.Hsapiens.UCSC.hg38.knownGene))
suppressMessages(library(clusterProfiler))
suppressMessages(library(ChIPseeker))
suppressMessages(library(DT))
suppressMessages(library(tidyverse))
suppressMessages(library(plotly))
suppressMessages(library(msigdbr))
suppressMessages(library(enrichR))
```

## 0) Load data

In this ChIP-Seq analysis the goal was analyze two ChIP-Seq data sets and compare the results.\

The two data sets chosen for this ChIP-Seq analysis consist of `PRDM1` and and `ZNF414`.\

Gene `PRDM1` encodes protein `PR domain zinc finger protein 1`/`B lymphocyte-induced maturation protein-1 (BLIMP-1)`. This protein is expressed in both B and T cells and plays a significant role in B cell development and antibody production.\

Gene `ZNF414` encodes protein `Zinc Finger Protein 414`. Its function is not clear but it may be involved in transcriptional regulation.\

These two data sets were chosen because an analysis of significant overlaps could enable more inside on how these two proteins may form a complex or have interaction in regulation chromosome remodelling or gene expression.\

The two data sets analyzed were downloaded from [ENCODE](https://www.encodeproject.org/):`PRDM1` ([data set](https://www.encodeproject.org/experiments/ENCSR977FEF/) -- [peaks](https://www.encodeproject.org/files/ENCFF719BHI/)), `ZNF414` ([data set](https://www.encodeproject.org/experiments/ENCSR224NQI/) -- [peaks](https://www.encodeproject.org/files/ENCFF409BJK/)).\ 

Both peaks for `PRDM1` and `ZNF414` were downloaded in `bed narrowPeak` format and read-in using R package `ChIPseeker` function `readPeakFile` for further analysis.\

Both peak files contained `IDR thresholded peaks` and used genome assembly `GRCh38`.

## 1) Venn diagram comparing the overlap of binding sites between PRDM1 and ZNF414

Both `PRDM1` and `ZNF414` showed a large number of peaks. The overlap between the two showed 2106 peaks.

```{r, results='hide'}
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

# ChIPseeker provides readPeakFile to load the peak and store in GRanges object
PRDM1 <- readPeakFile("PRDM1.bed")
ZNF414 <- readPeakFile("ZNF414.bed")

# Calculate the overlapping peaks between PRDM1 and ZNF414 with ChIPpeakAnno
ol <- findOverlapsOfPeaks(PRDM1, ZNF414)

# Visualize the overlapping using function makeVennDiagram from ChIPpeakAnno
makeVennDiagram(ol, fill=c("#009E73", "#F0E442"), # circle fill color
                col=c("#D55E00", "#0072B2"), #circle border color
                cat.col=c("#D55E00", "#0072B2")) # label color, keep same as circle border color
```

## 2) Metaplot of PRDM1 and ZNF414 around the transcription start sites (TSS)

### 2.1) Heatmap of ChIP binding to TSS regions

First, for calculating the profile of ChIP peaks binding to TSS regions, we prepared the TSS regions, which are defined as the flanking sequence of the TSS sites. Therefore, we aligned the peaks that mapped to these regions.\

`PRDM1///ZNF414` stands for the overlaps between `PRDM1` and `ZNF414`.

```{r}
promoter <- getPromoters(TxDb=txdb, upstream=3000, downstream=3000)

tagMatrix_PRDM1 <- getTagMatrix(PRDM1, windows=promoter)
tagMatrix_ZNF414 <- getTagMatrix(ZNF414, windows=promoter)

# Get overlapped peaks for PRDM1 and ZNF414
overlaps <- ol$peaklist[["PRDM1///ZNF414"]]

tagMatrix_overlaps <- getTagMatrix(overlaps, windows=promoter)

tagHeatmap(list(PRDM1=tagMatrix_PRDM1, ZNF414=tagMatrix_ZNF414, "PRDM1///ZNF414"=tagMatrix_overlaps), xlim=c(-3000, 3000), color="red")
```

### 2.2) Average Profile of ChIP peaks binding to TSS region

```{r}
plotAvgProf(list(PRDM1=tagMatrix_PRDM1, ZNF414=tagMatrix_ZNF414, "PRDM1///ZNF414"=tagMatrix_overlaps), xlim=c(-3000, 3000),
            xlab="Genomic Region (5'->3')", ylab = "Read Count Frequency")
```

## 3) Annotate the peaks for genomic features such as intron, exon, 3’UTR, etc and compare the annotations between between PRDM1 and ZNF414

First, we summarized the distribution of peaks `PRMD1` and `ZNF414` over different type of features such as exon, intron, enhancer, proximal promoter, 5’ UTR and 3’ UTR.

```{r}
peaks <- GRangesList(ZNF414=ZNF414,
                     PRDM1=PRDM1)

# Check the genomic element distribution of the duplicates
# The genomic element distribution will indicates the correlation between duplicates
genomicElementDistribution(peaks, 
                           TxDb = txdb,
                           promoterRegion=c(upstream=2000, downstream=500),
                           geneDownstream=c(upstream=0, downstream=2000))
```

We also checked the genomic element distribution for the overlaps.

```{r}
out <- genomicElementDistribution(overlaps, 
                           TxDb = txdb,
                           promoterRegion=c(upstream=2000, downstream=500),
                           geneDownstream=c(upstream=0, downstream=2000),
                           promoterLevel=list(
                         # from 5' -> 3', fixed precedence 3' -> 5'
                             breaks = c(-2000, -1000, -500, 0, 500),
                             labels = c("upstream 1-2Kb", "upstream 0.5-1Kb", 
                                        "upstream <500b", "TSS - 500b"),
                             colors = c("#FFE5CC", "#FFCA99", 
                                        "#FFAD65", "#FF8E32")))
```

We created upsetplots for `PRMD1`, `ZNF414` and the overlaps.

```{r}
# Assign genes to PRDM1 peaks
peakAnno_PRDM1 <- annotatePeak(PRDM1, tssRegion=c(-3000, 3000),
                         TxDb=txdb, annoDb="org.Hs.eg.db", verbose=FALSE)

upsetplot(peakAnno_PRDM1) + ggtitle("Upsetplot PRDM1")
```

```{r}
# Assign genes to ZNF414 peaks
peakAnno_ZNF414 <- annotatePeak(ZNF414, tssRegion=c(-3000, 3000),
                         TxDb=txdb, annoDb="org.Hs.eg.db", verbose=FALSE)

upsetplot(peakAnno_ZNF414)  + ggtitle("Upsetplot ZNF414")
```

```{r}
# Assign genes to overlaps peaks
peakAnno_overlaps <- annotatePeak(overlaps, tssRegion=c(-3000, 3000),
                         TxDb=txdb, annoDb="org.Hs.eg.db", verbose=FALSE)

upsetplot(peakAnno_overlaps)  + ggtitle("Upsetplot PRDM1///ZNF414")
```

We calculated the percentage of binding sites upstream and downstream from the TSS of the nearest genes and visualized the distribution.

```{r}
plotDistToTSS(list(PRDM1=peakAnno_PRDM1, ZNF414=peakAnno_ZNF414, "PRDM1///ZNF414"=peakAnno_overlaps), title="Distribution of transcription factor-binding loci \n relative to TSS")
```

## 4) Assign peaks to genes – then perform pathway enrichment.

For enrichment of the annotated peaks R package `clusterProfiler` function `enricher` was used. For the enrichment we used ontology gene sets of species Homo sapiens from R package `msigdbr`. Finally, the found pathways were filtered to contain only terms with an adjusted p-value smaller than or equal 0.05.

```{r}
# Ensure a deterministic result for enricher
set.seed(5)

# Get the gene sets
gene_sets <- msigdbr(species = "Homo sapiens", category = "C5") # C5: ontology gene sets
term2gene <- gene_sets[, c("gs_exact_source", "entrez_gene")]
term2name <- gene_sets[, c("gs_exact_source", "gs_name")]
```

```{r}
# Get annotation data frame
PRDM1_annot <- as.data.frame(peakAnno_PRDM1@anno)

# clusterProfiler provides enricher function for hypergeometric test 
# Performed pathway enrichment for PRDM1
PRDM1_enrichr <- enricher(PRDM1_annot$geneId,
                  pAdjustMethod = "BH",
                  qvalueCutoff = 0.05,
                  TERM2GENE=term2gene, 
                  TERM2NAME=term2name)

PRDM1_enrichrdf <- PRDM1_enrichr %>%
  filter(p.adjust <= 0.05) %>%
  arrange(desc(Count)) %>%
  select(Description, p.adjust, geneID, Count) %>%
  .@result
```

```{r}
# Get annotation data frame
ZNF414_annot <- as.data.frame(peakAnno_ZNF414@anno)

# Performed pathway enrichment for ZNF414
ZNF414_enrichr <- enricher(ZNF414_annot$geneId,
                  pAdjustMethod = "BH",
                  qvalueCutoff = 0.05,
                  TERM2GENE=term2gene, 
                  TERM2NAME=term2name)

ZNF414_enrichrdf <- ZNF414_enrichr %>%
  filter(p.adjust <= 0.05) %>%
  arrange(desc(Count)) %>%
  select(Description, p.adjust, geneID, Count) %>%
  .@result
```

```{r}
# Get annotation data frame
overlaps_annot <- as.data.frame(peakAnno_overlaps@anno)

# Performed pathway enrichment for overlaps
overlaps_enrichr <- enricher(overlaps_annot$geneId,
                  pAdjustMethod = "BH",
                  qvalueCutoff = 0.05,
                  TERM2GENE=term2gene, 
                  TERM2NAME=term2name)

overlaps_enrichrdf <- overlaps_enrichr %>%
  filter(p.adjust <= 0.05) %>%
  arrange(desc(Count)) %>%
  select(Description, p.adjust, geneID, Count) %>%
  .@result
```

### 4.1) What are genes shared in the overlap between PRDM1 and ZNF414? 

**Table 1** shows the identified 1819 genes shared in the overlap between PRDM1 and ZNF414.

```{r}
intersect_genes <- overlaps_annot %>%
  select(geneId, ENSEMBL, SYMBOL, GENENAME) %>%
  distinct(SYMBOL, .keep_all = TRUE)

datatable(intersect_genes, options=list(scrollX = TRUE, scrollY = TRUE), caption = 'Table 1: Genes in the overlap between PRDM1 and ZNF414', class = 'cell-border stripe')
```

### 4.2) What are pathways/genesets shared in the overlap between PRDM1 and ZNF414 for ontology gene sets from R package <code>msigdbr</code>? 

**Table 2** shows the 155 pathways/genesets shared in the overlap between PRDM1 and ZNF414.

```{r}
datatable(overlaps_enrichrdf, options=list(scrollX = TRUE, scrollY = TRUE), caption = 'Table 2: Pathways/genesets in the overlap between PRDM1 and ZNF414', class = 'cell-border stripe')
```

### 4.3) What pathways differ for ontology gene sets from R package <code>msigdbr</code>? 

**Table 3** shows the 37 pathways that differ between the overlap and PRDM1, i.e. pathways that are in the overlap but not in PRDM1.

```{r}
outersect_enrichr_PRDM1 <- overlaps_enrichrdf %>%
  rownames_to_column() %>%
  filter(!rowname %in% rownames(PRDM1_enrichrdf)) %>%
  arrange(desc(Count)) %>%
  select(-p.adjust, -Count)

outersect_enrichr_PRDM1 <- column_to_rownames(outersect_enrichr_PRDM1, var = "rowname")

datatable(outersect_enrichr_PRDM1, options=list(scrollX = TRUE, scrollY = TRUE), caption = 'Table 3: Pathways that differ between the overlap and PRDM1', class = 'cell-border stripe')
```

**Table 4** shows the 22 pathways that differ between the overlap and ZNF414, i.e. pathways that are in the overlap but not in ZNF414.

```{r}
outersect_enrichr_ZNF414 <- overlaps_enrichrdf %>%
  rownames_to_column() %>%
  filter(!rowname %in% rownames(ZNF414_enrichrdf)) %>%
  arrange(desc(Count)) %>%
  select(-p.adjust, -Count)

outersect_enrichr_ZNF414 <- column_to_rownames(outersect_enrichr_ZNF414, var = "rowname")

datatable(outersect_enrichr_ZNF414, options=list(scrollX = TRUE, scrollY = TRUE), caption = 'Table 4: Pathways that differ between the overlap and ZNF414', class = 'cell-border stripe')
```

### 5) Enrichr

Additionally to the ontology gene sets from R package `msigdbr`, we performed another gene enrichment analysis. Gene ontology is a very limited pathway annotation that is hard to interpret for many applications. Therefore, we also performed a broad-spectrum [enrichr](https://maayanlab.cloud/Enrichr/) pathway analysis.\

```{r}
# Save found 1819 overlap genes for upload to enrichr
#write.table(unique(overlaps_annot$SYMBOL),"overlaps_genes.txt",sep="\t",quote=FALSE,row.names=FALSE, col.names=FALSE)
```

The enrichr results for the 1819 overlap genes can be found here:
https://maayanlab.cloud/Enrichr/enrich?dataset=a64a28f137dea51300456970b16d9296

Subsequently, we had a closer look at our results for enrichr. Most found pathways belonged to databases related to transcription factors. Therefore, we have a closer look at enrichr database `Enrichr_Submissions_TF-Gene_Coocurrence` providing interesting results.\

The found pathways for this database were filtered to contain only terms with an adjusted p-value smaller than or equal 0.05 and a combined score greater 100.\

The formula for the combined score (c) is `ln(p) * z`. Variable `p` is the p-value computed using Fisher's exact test and variable `z` is the z-score computed to assess the deviation from the expected rank.

```{r, results='hide'}
setEnrichrSite("Enrichr") # Human genes
websiteLive <- TRUE
# dbs <- listEnrichrDbs() %>% select(libraryName) %>% unlist()
dbs <- "Enrichr_Submissions_TF-Gene_Coocurrence"
```

```{r, results='hide'}
# Performed pathway enrichment for PRDM1
enriched <- enrichr(PRDM1_annot$SYMBOL, dbs)

# Filtered significant pathways
enriched <- lapply(enriched, function(x) {
  x %>%
    filter(Adjusted.P.value <= 0.05 & Combined.Score > 100) 
})

PRDM1_enricheddf <- bind_rows(enriched, .id = 'Database') %>%
  mutate(id=paste0(Database, "-", Term)) %>%
  select(-Old.P.value, -Old.Adjusted.P.value) %>%
  arrange(desc(Combined.Score))
```

```{r, results='hide'}
# Performed pathway enrichment for ZNF414
enriched <- enrichr(ZNF414_annot$SYMBOL, dbs)

# Filtered significant pathways
enriched <- lapply(enriched, function(x) {
  x %>%
    filter(Adjusted.P.value <= 0.05 & Combined.Score > 100) 
})

ZNF414_enricheddf <- bind_rows(enriched, .id = 'Database') %>%
  mutate(id=paste0(Database, "-", Term)) %>%
  select(-Old.P.value, -Old.Adjusted.P.value) %>%
  arrange(desc(Combined.Score))
```

```{r, results='hide'}
# Performed pathway enrichment for overlaps
enriched <- enrichr(overlaps_annot$SYMBOL, dbs)

# Filtered significant pathways
enriched <- lapply(enriched, function(x) {
  x %>%
    filter(Adjusted.P.value <= 0.05 & Combined.Score > 100) 
})

overlaps_enricheddf <- bind_rows(enriched, .id = 'Database') %>%
  mutate(id=paste0(Database, "-", Term)) %>%
  select(-Old.P.value, -Old.Adjusted.P.value) %>%
  arrange(desc(Combined.Score))
```

### 5.1) What are pathways/genesets shared in the overlap between PRDM1 and ZNF414 for database <code>Enrichr_Submissions_TF-Gene_Coocurrence</code> from <code>enrichr</code>?

```{r}
my.labels <- overlaps_enricheddf$Term

overlaps_enricheddf$Term <- factor(overlaps_enricheddf$Term, levels = overlaps_enricheddf$Term[order(overlaps_enricheddf$Combined.Score)])

viz <- head(overlaps_enricheddf, n=10) %>%
  ggplot(aes(x=Term, y=Combined.Score, fill=Adjusted.P.value)) + 
  geom_bar(stat = "identity") + 
  ggtitle("Top 10 pathways selected database") +
  theme(plot.title = element_text(hjust = 0.5)) +
  xlab("") +
  scale_x_discrete(labels = function(x) str_wrap(x, width = 30)) + 
  coord_flip()

ggplotly(viz)
```

**Table 5** shows the 372 pathways/genesets shared in the overlap between PRDM1 and ZNF414.

```{r}
datatable(overlaps_enricheddf %>% select(-id), options=list(scrollX = TRUE, scrollY = TRUE), caption = 'Table 5: Pathways/genesets in the overlap between PRDM1 and ZNF414', class = 'cell-border stripe')
```

### 5.2) What pathways differ for database <code>Enrichr_Submissions_TF-Gene_Coocurrence</code> from <code>enrichr</code>?

**Table 6** shows the 95 pathways that differ between the overlap and PRDM1, i.e. pathways that are in the overlap but not in PRDM1.

```{r}
outersect_enriched_PRDM1 <- overlaps_enricheddf %>%
  filter(!id %in% PRDM1_enricheddf$id) %>%
  arrange(desc(Combined.Score)) %>%
  select(-id)

datatable(outersect_enriched_PRDM1, options=list(scrollX = TRUE, scrollY = TRUE), caption = 'Table 6: Pathways that differ between the overlap and PRDM1', class = 'cell-border stripe')
```

**Table 7** shows the 143 pathways that differ between the overlap and ZNF414, i.e. pathways that are in the overlap but not in ZNF414.

```{r}
outersect_enriched_ZNF414 <- overlaps_enricheddf %>%
  filter(!id %in% ZNF414_enricheddf$id) %>%
  arrange(desc(Combined.Score)) %>%
  select(-id)

datatable(outersect_enriched_ZNF414, options=list(scrollX = TRUE, scrollY = TRUE), caption = 'Table 7: Pathways that differ between the overlap and ZNF414', class = 'cell-border stripe')
```

## 6.1) What is your interpretation of these results? 

The goal of this ChIP-Seq analysis was analyze ChIP-Seq data sets `PRDM1` and `ZNF414`.\

As said before, gene `PRDM1` encodes protein `PR domain zinc finger protein 1`/`B lymphocyte-induced maturation protein-1 (BLIMP-1)`, which plays a significant role in B cell development and antibody production and gene `ZNF414` encodes protein `Zinc Finger Protein 414`, which function is not clear but may be involved in transcriptional regulation.\

These two data sets were chosen because an analysis of significant overlaps could enable more inside on how these two proteins may form a complex or have interaction in regulation chromosome remodelling or gene expression.\

As show in the venn diagram, we found 2106 peaks in the overlap between `PRDM1` and `ZNF414`.\

The genomic element distribution for the overlaps showed that on the exon level, 53.3% belonged to 5' UTR, on the exon/intron/intergenic level 68.1% belonged to exon, on the gene level 53.9% belonged to promotor and on the promotor level 46.6% belonged to TSS - 500b.\

Overall, our results show that the identified overlaps between `PRDM1` and `ZNF414` belonged to coding regions and more specifically, to TSS, which is an interesting result considering the unknown interactions in regulation chromosome remodelling or gene expression between the two.\

We then have a look at the top 5 pathway after our gene set enrichment analysis for the overlaps between `PRDM1` and `ZNF414`.\

We start with the identified pathways of ontology gene sets from R package `msigdbr`:\

We can see that two of them belong to RNA processes:

- [RNA catabolic process](https://www.ebi.ac.uk/QuickGO/term/GO:0006401)\
- [ribonucleoprotein complex biogenesis](https://www.ebi.ac.uk/QuickGO/term/GO:0022613)


One belongs to the nuclear cellular component nuclear envelope:

- [nuclear envelope](https://www.ebi.ac.uk/QuickGO/term/GO:0005635)


And two belong to ncRNA processes:

- [ncRNA metabolic process](https://www.ebi.ac.uk/QuickGO/term/GO:0034660)\
- [ncRNA processing](https://www.ebi.ac.uk/QuickGO/term/GO:0034470)

This could be an indicator that the identified overlaps between `PRDM1` and `ZNF414` consist of regions with genes involved in transcription and seems to confirm the assumption that `ZNF414` may be involved in transcriptional regulation.\

This assumption could also further be validated by our pathways that differ between the overlaps of `PRDM1` and `ZNF414`.\

The different pathways between the overlaps and `PRDM1` seem to mainly relate to chromatin / chromosome / transcriptional processes, e.g. [covalent chromatin modification](https://www.ebi.ac.uk/QuickGO/term/GO:0016569), [regulation of chromosome organization](https://www.ebi.ac.uk/QuickGO/term/GO:0033044) and [ribosome](https://www.ebi.ac.uk/QuickGO/term/GO:0005840).\

The different pathways between the overlaps and `ZNF414` showed partially similar results, like e.g. [regulation of posttranscriptional gene silencing](https://www.ebi.ac.uk/QuickGO/GTerm?id=GO:0060147) but also unrelated terms like e.g. [abnormal uterus morphology](http://www.gsea-msigdb.org/gsea/msigdb/geneset_page.jsp?geneSetName=HP_ABNORMAL_UTERUS_MORPHOLOGY).\

Overall, most of the identified pathways in the overlaps between `PRDM1` and `ZNF414` seem to be related to transcriptional regulation.\

As stated before, most identified pathways for the overlaps between `PRDM1` and `ZNF414` with broad-spectrum enrichr pathway analysis belonged to databases related to transcription factors, like enrichr databases `Enrichr_Submissions_TF-Gene_Coocurrence`,  `ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X` and `ENCODE TF ChIP-seq 2015`.\

`Enrichr_Submissions_TF-Gene_Coocurrence` contains 1722 terms, a gene coverage of 12486 genes and	299 genes per term.\

**TF** stands for transcription factor. In molecular biology, a TF is a protein that is important for the initiation of RNA polymerase during transcription. **Co-occurence** networks are used to describe potential relationships between entities like in this case TFs.\

Samples from the top 10 `Enrichr_Submissions_TF-Gene_Coocurrence` pathways are:

- [MNT](https://www.genecards.org/cgi-bin/carddisp.pl?gene=MNT) (MAX Network Transcriptional Repressor) is a Protein Coding gene. GO annotations related to this gene include DNA-binding transcription factor activity and transcription co-activator activity\
- [ZNF839](https://www.genecards.org/cgi-bin/carddisp.pl?gene=ZNF839) (Zinc Finger Protein 839) is a Protein Coding gene. Among its related pathways are Gene Expression\
- [MAFG](https://www.genecards.org/cgi-bin/carddisp.pl?gene=MAFG) (MAF BZIP Transcription Factor G)  is a Protein Coding gene. GO annotations related to this gene include DNA-binding transcription factor activity and protein heterodimerization activity.

Overall, our results strongly indicate that the identified overlaps between `PRDM1` and `ZNF414` are related to transcriptional regulation.

## 6.2) What future directions could you propose to follow up on these findings?

One direction for future research would be how these overlaps between `PRDM1` and `ZNF414` are involved in transcriptional regulation play a role in B cell development and antibody production. `ZNF414` may be more important for the regulation of `PRDM1` than previously thought.

