---
title: "ChIP_Seq_Analysis"
author: "Maximilian Greil"
date: "21 4 2021"
output:
  html_document:
    theme: journal
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

```{r}
suppressMessages(library(UpSetR))
suppressMessages(library(ChIPpeakAnno))
suppressMessages(library(TxDb.Hsapiens.UCSC.hg38.knownGene))
suppressMessages(library(org.Hs.eg.db))
suppressMessages(library(clusterProfiler))
suppressMessages(library(ChIPseeker))
suppressMessages(library(DT))
suppressMessages(library(annotables))
suppressMessages(library(tidyverse))
```

## 0) Load data

In this ChIP-Seq analysis the goal was analyze two ChIP-Seq data sets and compare the results.\

The two data sets chosen for this ChIP-Seq analysis consist of `PRDM1` and and `ZNF414`.\

Gene `PRDM1` encodes protein `PR domain zinc finger protein 1`/`B lymphocyte-induced maturation protein-1 (BLIMP-1)`. This protein is expressed in both B and T cells and plays a significant role in B cell development and antibody production.\

Gene `ZNF414` encodes protein `Zinc Finger Protein 414`. Its function is not clear but it may be involved in transcriptional regulation.\

These two data sets were chosen because an analysis of significant overlaps could enable more inside on how these two proteins may form a complex or have interaction in regulation chromosome remodelling or gene expression.\

The two data sets analyzed were downloaded from [ENCODE](https://www.encodeproject.org/):`PRDM1` ([data set](https://www.encodeproject.org/experiments/ENCSR977FEF/) -- [peaks](https://www.encodeproject.org/files/ENCFF719BHI/)), `ZNF414` ([data set](https://www.encodeproject.org/experiments/ENCSR224NQI/) -- [peaks](https://www.encodeproject.org/files/ENCFF409BJK/)).\ 

Both peaks for `PRDM1` and `ZNF414` were downloaded in `bed narrowPeak` format and read-in using R package `ChIPseeker` function `readPeakFile` for further analysis.\

Both peak files contained `IDR thresholded peaks` and used genome assembly `GRCh38`.

## 1) Venn diagram comparing the overlap of binding sites between PRDM1 and ZNF414

Both `PRDM1` and `ZNF414` showed a large number of peaks. The overlap between the two showed 2106 peaks.

```{r, results='hide'}
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

# ChIPseeker provides readPeakFile to load the peak and store in GRanges object
PRDM1 <- readPeakFile("PRDM1.bed")
ZNF414 <- readPeakFile("ZNF414.bed")

# Calculate the overlapping peaks between PRDM1 and ZNF414 with ChIPpeakAnno
ol <- findOverlapsOfPeaks(PRDM1, ZNF414)

# Visualize the overlapping using function makeVennDiagram from ChIPpeakAnno
makeVennDiagram(ol, fill=c("#009E73", "#F0E442"), # circle fill color
                col=c("#D55E00", "#0072B2"), #circle border color
                cat.col=c("#D55E00", "#0072B2")) # label color, keep same as circle border color
```

## 2) Metaplot of PRDM1 and ZNF414 around the transcription start sites (TSS)

### 2.1) Heatmap of ChIP binding to TSS regions

First, for calculating the profile of ChIP peaks binding to TSS regions, we prepared the TSS regions, which are defined as the flanking sequence of the TSS sites. Therefore, we aligned the peaks that mapped to these regions.

```{r}
promoter <- getPromoters(TxDb=txdb, upstream=3000, downstream=3000)

tagMatrix_PRDM1 <- getTagMatrix(PRDM1, windows=promoter)
tagMatrix_ZNF414 <- getTagMatrix(ZNF414, windows=promoter)

tagHeatmap(list(PRDM1=tagMatrix_PRDM1, ZNF414=tagMatrix_ZNF414), xlim=c(-3000, 3000), color="red")
```

### 2.2) Average Profile of ChIP peaks binding to TSS region

```{r}
plotAvgProf(list(PRDM1=tagMatrix_PRDM1, ZNF414=tagMatrix_ZNF414), xlim=c(-3000, 3000),
            xlab="Genomic Region (5'->3')", ylab = "Read Count Frequency")
```

### 2.3) Plot distribution of the distance of overlapped peaks between PRDM1 and ZNF414 to TSS

```{r}
# Annotation data should be an object of GRanges.
annoData <- toGRanges(txdb, feature="gene")

# Get overlapped peaks for PRDM1 and ZNF414
overlaps <- ol$peaklist[["PRDM1///ZNF414"]]

# Plot distribution of the distance of overlapped peaks to transcription start sites (TSS)
binOverFeature(overlaps, annotationData=annoData,
               radius=3000, FUN=length, errFun=0,
               xlab="distance from TSS (bp)", ylab="count", 
               main="Distribution of aggregated peak numbers around TSS")
```

## 3) Annotate the peaks for genomic features such as intron, exon, 3’UTR, etc and compare the annotations between between PRDM1 and ZNF414

First, we summarized the distribution of peaks `PRMD1` and `ZNF414` over different type of features such as exon, intron, enhancer, proximal promoter, 5’ UTR and 3’ UTR.

```{r}
peaks <- GRangesList(ZNF414=ZNF414,
                     PRDM1=PRDM1)

# Check the genomic element distribution of the duplicates
# The genomic element distribution will indicates the correlation between duplicates
genomicElementDistribution(peaks, 
                           TxDb = txdb,
                           promoterRegion=c(upstream=2000, downstream=500),
                           geneDownstream=c(upstream=0, downstream=2000))
```

We created upsetplots for `PRMD1` and `ZNF414`.

```{r}
# Assign genes to PRDM1 peaks
peakAnno_PRDM1 <- annotatePeak(PRDM1, tssRegion=c(-3000, 3000),
                         TxDb=txdb, annoDb="org.Hs.eg.db", verbose=FALSE)

upsetplot(peakAnno_PRDM1) + ggtitle("Upsetplot PRDM1")
```

```{r}
# Assign genes to ZNF414 peaks
peakAnno_ZNF414 <- annotatePeak(ZNF414, tssRegion=c(-3000, 3000),
                         TxDb=txdb, annoDb="org.Hs.eg.db", verbose=FALSE)

upsetplot(peakAnno_ZNF414)  + ggtitle("Upsetplot ZNF414")
```

We calculated the percentage of binding sites upstream and downstream from the TSS of the nearest genes and visualized the distribution.

```{r}
plotDistToTSS(list(PRDM1=peakAnno_PRDM1, ZNF414=peakAnno_ZNF414), title="Distribution of transcription factor-binding loci \n relative to TSS")
```

We also checked the genomic element distribution for the overlaps.

```{r}
out <- genomicElementDistribution(overlaps, 
                           TxDb = txdb,
                           promoterRegion=c(upstream=2000, downstream=500),
                           geneDownstream=c(upstream=0, downstream=2000),
                           promoterLevel=list(
                         # from 5' -> 3', fixed precedence 3' -> 5'
                             breaks = c(-2000, -1000, -500, 0, 500),
                             labels = c("upstream 1-2Kb", "upstream 0.5-1Kb", 
                                        "upstream <500b", "TSS - 500b"),
                             colors = c("#FFE5CC", "#FFCA99", 
                                        "#FFAD65", "#FF8E32")))
```

## 4) Assign peaks to genes – then perform pathway enrichment.

```{r}
# Get annotation data frame
PRDM1_annot <- as.data.frame(peakAnno_PRDM1@anno)

# Performed pathway enrichment for PRDM1
PRDM1_ego <- enrichGO(gene = PRDM1_annot$geneId, 
                    keyType = "ENTREZID", 
                    OrgDb = org.Hs.eg.db, 
                    ont = "BP", # Biological Process 
                    pAdjustMethod = "BH", 
                    qvalueCutoff = 0.05, 
                    readable = TRUE)

PRDM1_egodf <- PRDM1_ego %>%
  filter(p.adjust <= 0.05) %>%
  arrange(desc(Count)) %>%
  select(Description, p.adjust, geneID, Count) %>%
  .@result

```

```{r}
# Get annotation data frame
ZNF414_annot <- as.data.frame(peakAnno_ZNF414@anno)

# Performed pathway enrichment for ZNF414
ZNF414_ego <- enrichGO(gene = ZNF414_annot$geneId, 
                    keyType = "ENTREZID", 
                    OrgDb = org.Hs.eg.db, 
                    ont = "BP", # Biological Process 
                    pAdjustMethod = "BH", 
                    qvalueCutoff = 0.05, 
                    readable = TRUE)

ZNF414_egodf <- ZNF414_ego %>%
  filter(p.adjust <= 0.05) %>%
  arrange(desc(Count)) %>%
  select(Description, p.adjust, geneID, Count) %>%
  .@result

```

```{r}
# Assign genes to overlaps peaks
peakAnno_overlaps <- annotatePeak(overlaps, tssRegion=c(-3000, 3000),
                         TxDb=txdb, annoDb="org.Hs.eg.db", verbose=FALSE)

# Get annotation data frame
overlaps_annot <- as.data.frame(peakAnno_overlaps@anno)

# Performed pathway enrichment for overlaps
overlaps_ego <- enrichGO(gene = overlaps_annot$geneId, 
                    keyType = "ENTREZID", 
                    OrgDb = org.Hs.eg.db, 
                    ont = "BP", # Biological Process 
                    pAdjustMethod = "BH", 
                    qvalueCutoff = 0.05, 
                    readable = TRUE)

overlaps_egodf <- overlaps_ego %>%
  filter(p.adjust <= 0.05) %>%
  arrange(desc(Count)) %>%
  select(Description, p.adjust, geneID, Count) %>%
  .@result
```

### 4.1) What are genes and pathways/genesets shared in the overlap between PRDM1 and ZNF414? 

**Table 1** shows the identified 1853 genes shared in the overlap between PRDM1 and ZNF414.

```{r}
intersect_genes <- grch38 %>% 
  filter(entrez %in% overlaps_annot$geneId) %>%
  select(-chr, -strand, -biotype)

datatable(intersect_genes, options=list(scrollX = TRUE, scrollY = TRUE), caption = 'Table 1: Genes in the overlap between PRDM1 and ZNF414', class = 'cell-border stripe')
```

**Table 2** shows the 131 pathways/genesets shared in the overlap between PRDM1 and ZNF414.

```{r}
datatable(overlaps_egodf, options=list(scrollX = TRUE, scrollY = TRUE), caption = 'Table 2: Pathways/genesets in the overlap between PRDM1 and ZNF414', class = 'cell-border stripe')
```

### 4.2) What pathways differ? 

**Table 3** shows the 24 pathways that differ between the overlap and PRDM1, i.e. pathways that are in the overlap but not in PRDM1.

```{r}
outersect_ego_PRDM1 <- overlaps_egodf %>%
  rownames_to_column() %>%
  filter(!rowname %in% rownames(PRDM1_egodf)) %>%
  arrange(desc(Count)) %>%
  select(-p.adjust, -Count)

outersect_ego_PRDM1 <- column_to_rownames(outersect_ego_PRDM1, var = "rowname")

datatable(outersect_ego_PRDM1, options=list(scrollX = TRUE, scrollY = TRUE), caption = 'Table 3: Pathways that differ between the overlap and PRDM1', class = 'cell-border stripe')
```

**Table 4** shows the 12 pathways that differ between the overlap and ZNF414, i.e. pathways that are in the overlap but not in ZNF414.

```{r}
outersect_ego_ZNF414 <- overlaps_egodf %>%
  rownames_to_column() %>%
  filter(!rowname %in% rownames(ZNF414_egodf)) %>%
  arrange(desc(Count)) %>%
  select(-p.adjust, -Count)

outersect_ego_ZNF414 <- column_to_rownames(outersect_ego_ZNF414, var = "rowname")

datatable(outersect_ego_ZNF414, options=list(scrollX = TRUE, scrollY = TRUE), caption = 'Table 4: Pathways that differ between the overlap and ZNF414', class = 'cell-border stripe')
```


## 5.1) What is your interpretation of these results? 

The goal of this ChIP-Seq analysis was analyze ChIP-Seq data sets `PRDM1` and `ZNF414`.\

As said before, gene `PRDM1` encodes protein `PR domain zinc finger protein 1`/`B lymphocyte-induced maturation protein-1 (BLIMP-1)`, which plays a significant role in B cell development and antibody production and gene `ZNF414` encodes protein `Zinc Finger Protein 414`, which function is not clear but may be involved in transcriptional regulation.\

These two data sets were chosen because an analysis of significant overlaps could enable more inside on how these two proteins may form a complex or have interaction in regulation chromosome remodelling or gene expression.\

As show in the venn diagram, we found 2106 peaks in the overlap between `PRDM1` and `ZNF414`.\

The genomic element distribution for the overlaps showed that on the exon level, 53.3% belonged to 5' UTR, on the exon/intron/intergenic level 68.1% belonged to exon, on the gene level 53.9% belonged to promotor and on the promotor level 46.6% belonged to TSS - 500b.\

Overall, this result shows that the identified overlaps between `PRDM1` and `ZNF414` belonged to coding regions and more specifically, to TSS, which is an interesting result considering the unknown interactions in regulation chromosome remodelling or gene expression between the two.\

When we have a look at the top 5 pathway after our gene set enrichment analysis for the overlaps, we can see that they belong to RNA processes, e.g.:

- [ribonucleoprotein complex biogenesis](https://www.ebi.ac.uk/QuickGO/term/GO:0022613)
- [RNA catabolic process](https://www.ebi.ac.uk/QuickGO/term/GO:0006401)
- [RNA splicing](https://www.ebi.ac.uk/QuickGO/term/GO:0008380)

This could be an indicator that the identified overlaps between `PRDM1` and `ZNF414` consist of regions with genes involved in transcription and seems to confirm the assumption that `ZNF414` may be involved in transcriptional regulation. This assumption could also further be validated by our pathways that differ between the overlaps and `PRDM1` and `ZNF414`. The identified different pathways all seem to be related to transcriptional regulation.

## 5.2) What future directions could you propose to follow up on these findings?

One direction for future research would be how these overlaps between `PRDM1` and `ZNF414` are involved in transcriptional regulation play a role in B cell development and antibody production. `ZNF414` may be more important for the regulation of `PRDM1` than previously thought.

