---
title: "RNA_Seq_Analysis"
author: "Maximilian Greil"
date: "11 4 2021"
output:
  html_document:
    toc: true
    theme: journal
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries

```{r}
#suppressMessages(library(recount))
suppressMessages(library(DT))
suppressMessages(library(DESeq2))
suppressMessages(library(tidyverse))
suppressMessages(library(AnnotationDbi))
suppressMessages(library(org.Hs.eg.db))
suppressMessages(library(PoiClaClu))
suppressMessages(library(pheatmap))
suppressMessages(library(RColorBrewer))
suppressMessages(library(EnhancedVolcano))
suppressMessages(library(msigdbr))
suppressMessages(library(clusterProfiler))
```

### 0) Load and prepare data

First look at data

```{r}
dat <- read.csv('SRP008552.txt', header = TRUE, sep = "\t")

datatable(dat, class = 'cell-border stripe')
```

Load data

```{r}
## Comprehensive microRNA profiling in B-cells of human centenarians by massively parallel sequencing
## https://doi.org/10.1186/1471-2164-13-353

## https://www.ncbi.nlm.nih.gov/Traces/study/?acc=SRP008552&search=RNA-seq&o=acc_s%3Aa

## Data was dowloaded using package recount
# url <- download_study('SRP008552')

## Load the RangedSummarizedExperiment object
load(file.path('SRP008552', 'rse_gene.Rdata'))
```

Create additional column 'condition' for later creation of design matrix used for differential expression analysis. The two levels in 'condition' are 'control' and 'centenarian' based on column 'title'.

```{r}
## Create additional column for design matrix based on column title
rse_gene$condition <- factor(c("control", "control", "control", "centenarian", "centenarian", "centenarian"), levels = c("centenarian","control")) %>% relevel(.$condition, ref = "control")

## Examine data
rse_gene
```

Pre-filtering the dataset

```{r}
## Create DESeq data set, using paired design
dds <- DESeqDataSet(rse_gene, design = ~ condition)

## Remove any reads with a copy number less than 10
dds <- dds[rowSums(counts(dds)) >= 10,]

## Remove any reads observed in fewer than 3 samples
dds <- dds[rowSums(counts(dds) == 0) <= 3,]
```

## 1) Exploratory analyis and visualization

### 1.1) Create a PCA plot colored by the condition of interest

```{r}
## Perform regularized-logarithm transformation (rlog) on the data
rld <- rlog(dds)

plotPCA(rld, intgroup = c("condition", "title"))
```

### 1.2) Heatmap of Poisson distances between samples

```{r}
poisd <- PoissonDistance(t(counts(dds)))
samplePoisDistMatrix <- as.matrix( poisd$dd )
rownames(samplePoisDistMatrix) <- paste( dds$condition, dds$title, sep=" - " )
colnames(samplePoisDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         col = colors)
```

We can see in both the PCA plot and the Heatmap that samples for Centenarian 1 and Control 2 are more distant to their respective counterparts. Interestingly, the others are closer to each other.

## 2) Differential expression analysis

### 2.1) Use DESeq2 for the differential gene expression analysis. Find the DEGs between your conditions of interest.

Perform differential expression analysis

```{r}
res <- DESeq(dds) %>%
  results(.)

## Examine data
summary(res)
```

Annotating results

```{r}
## Filter DEGs for significant ones
idx <- which(res$padj < 0.05 & 
               abs(res$log2FoldChange) >= 1 & 
               res$baseMean >= 20 )

sigRes <- res[idx, ]

## Map DEGs to symbol and entrez
ens.str <- substr(rownames(sigRes), 1, 15)

sigRes$symbol <- mapIds(org.Hs.eg.db,
                     keys=ens.str,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first") %>%
  unname

sigRes$entrez <- mapIds(org.Hs.eg.db,
                     keys=ens.str,
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first") %>%
  unname

## Replaced NAs in columns symbol and entrez with string
sigRes[which( is.na( sigRes$symbol ) ), "symbol"] = "NA"
sigRes[which( is.na( sigRes$entrez ) ), "entrez"] = "NA"

## Ordered DEGs from strongest downregulation to upregulation
datatable(as.data.frame(sigRes[order(sigRes$log2FoldChange),]), class = 'cell-border stripe')
```

## 3) Plotting results

### 3.1) MA plot

```{r}
plotMA(res)
```

### 3.2) Create a Volcano Plot

```{r, fig.height=8.5, fig.width=7.5}
## Volcano plot for significant DEGs
EnhancedVolcano(sigRes, 
                lab = sigRes$symbol,
                x = "log2FoldChange", 
                y = "padj", 
                title = "Control versus Centenarian")
```

## 4) Create a heatmap showing the top 20 over- and under-expressed DEGs

### 4.1) Prepare matrix for heatmap

```{r}
mat <- assay(rld)
orderedSig <- sigRes[order(sigRes$padj), ]
id1 <- rownames(orderedSig)
id2 <- orderedSig$symbol
DE <- mat[id1,]
rownames(DE) <- id2
```

### 4.2) Plot heatmap of top 20 DEGs

```{r}
top20DE <- head(DE, n=20)
DEgenesTop <- mat[top20DE,]
annotation <- as.data.frame(colData(rld)[, c("condition","title")])
pheatmap(top20DE, scale = "row", clustering_distance_rows = "correlation", annotation_col = annotation, main="Top 20 DEGs")
```

## 5) Do GSEA on the results and plot the top 5 pathways

### 5.1) Prepare data for GSEA

```{r}
# Get the gene sets and wrangle
gene_sets <- msigdbr(species = "Homo sapiens", category = "C5")
gene_sets <- gene_sets %>%
  dplyr::select(gs_name, gene_symbol)

sigResdf <- as.data.frame(sigRes)

## Adding a score for GSEA
sigResdf2 <- sigResdf %>%
  arrange(padj) %>%
  mutate(gsea_metric = -log10(padj) * sign(log2FoldChange)) %>%
  arrange(desc(gsea_metric))

## GSEA value histogram
hist(sigResdf2$gsea_metric, breaks = 100)

## Get the ranked GSEA vector
ranks <- sigResdf2 %>%
  dplyr::select(symbol, gsea_metric) %>%
  dplyr::distinct(symbol, .keep_all = TRUE) %>%
  deframe()
```

### 5.2) Perform GSEA

```{r}
# Run GSEA
gseares <- GSEA(geneList = ranks, 
                TERM2GENE = gene_sets)
```

There was no enriched term found for p value Cutoff < 0.05. This is not necessarily bad, as the subject of this study is senescence, a research field of highly unknown specifics.

## 6) Concluding questions

### 6.1) What do the results tell you about your conditions of interest?

The data from this RNA-Seq analysis was generated by sequencing B-cells of human centenarians and controls.\

Therefore, the conditions of interest was the expected differential gene expression between centenarians and controls, i.e. which genes play an active role in human senescence.\

The results show that there are indeed DEGs between centenarians and controls.\

Human senescence seems to be correlated with specific gene expression patterns. Finding these genetic patterns and decrypting related gene pathways could have a big impact in solving the question of human senescence.\

### 6.2) What is the biological meaning of these results?

In this RNA-Seq analysis we found that a total of 27 genes showed differential expression between centenarians and controls. 15 of these 27 DEGs were upregulated and 12 were downregulated. These found genes can have an impact on human senescence.\

However, during GSEA no enriched terms were found. As explained before, this result is not necessarily bad because it can lead to new gene pathways correlated to human senescence.

From a biological perspective it is an interesting question why centenarians show this different gene expression, i.e. why are these genes differentially expressed for the controls?

### 6.3) What future experiments could you perform?

Future experiments could involve the analysis of the found genes in conext of GSEA, i.e. if the found genes could be involved in gene pathways related to human senescence.\

The experiment could be repeated but this time with a much larger sample size. This could be helpful in distinguishing important senescence genes between true - and false positives found in this analysis.